rvps/src/reference_value.rs

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) 2022 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+//! reference value for RVPS
+
+use chrono::{DateTime, NaiveDateTime, Utc};
+use serde::{Deserialize, Deserializer, Serialize};
+
+/// Default version of ReferenceValue
+pub const REFERENCE_VALUE_VERSION: &str = "0.1";
+
+/// A HashValuePair stores a hash algorithm name
+/// and relative artifact's hash value due to
+/// the algorithm.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
+pub struct HashValuePair {
+    alg: String,
+    value: String,
+}
+
+impl HashValuePair {
+    pub fn new(alg: String, value: String) -> Self {
+        Self { alg, value }
+    }
+
+    pub fn alg(&self) -> &String {
+        &self.alg
+    }
+
+    pub fn value(&self) -> &String {
+        &self.value
+    }
+}
+
+/// Helper to deserialize an expired time
+fn primitive_date_time_from_str<'de, D: Deserializer<'de>>(
+    d: D,
+) -> Result<DateTime<Utc>, D::Error> {
+    let s = <Option<&str>>::deserialize(d)?
+        .ok_or_else(|| serde::de::Error::invalid_length(0, &"<TIME>"))?;
+
+    let ndt = NaiveDateTime::parse_from_str(s, "%Y-%m-%dT%H:%M:%SZ")
+        .map_err(|err| serde::de::Error::custom::<String>(err.to_string()))?;
+
+    Ok(DateTime::<Utc>::from_utc(ndt, Utc))
+}
+
+/// Define Reference Value.
+/// This Reference Value is not the same as Reference in IETF's RATS.
+/// Here, RV is consumed by AS. Its format MAY be modified often to
+/// cowork with AS.
+/// * `version`: version of the reference value format.
+/// * `name`: name of the artifact related to this reference value.
+/// * `expired`: expired time for this reference value.
+/// * `hash_value`: A set of key-value pairs, each indicates a hash
+/// algorithm and its relative hash value for the artifact.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
+pub struct ReferenceValue {
```

Comment:
A little suggestion: I'm not sure if we really need such a heavy RV structure, any fields here may be unnecessary, such as `version`, `expired` and the `algo` in `HashValuePair`. 

For `version`, since this structure is not transmitted in cross process communication, it only needs to be consistent with the AS version. It is unnecessary to list it separately here.

For `expired`, I guess there will be a so-called "subscription" mechanism in the AS to update the reference value from the upstream in time. For the caller, the latest reference value will be obtained from the rvps every time the comparison verification is performed. Therefore, the 'expired' here may not be used.

For `algo`, because the reference value provided by rvps is mainly used for reference, and in fact we do not know what algorithm is used for the compared real value. In comparison and verification, we only need to ensure that they are the same. Therefore, even if there are reference values of multiple algorithms, we only need to simply compare whether the real value is equal to one of them. Therefore, it is not necessary to involve what algorithm they use.

So, perhaps a simple RV structural design can meet all requirements:

```rust
ReferenceValue {
    // The resource name of this RV.
    name: String,
    // The reference hash values, base64 coded.
    hash_values: Vec<String>,
}
```

I'm not sure if this is the best, if I missed something, please remind me :)

Reply:
Thanks for your deep dive into the code!

Certainly we can delete these fields, however:
* `version` may give some extensibility for possible changes in the future
* Although `expired` can give an expired time because provenance information like `in-toto` will declare "when this metadata will be expired" without using the publishing mechanism you've mentioned, I think it is more suitable to handle inside RVPS.
*  `algo` is the most controversial field, I preferred to record the digest algorithm for potential usage and extensibility. Maybe it's wiser to delete this for current implementation. In the future, perhaps, we can design a new format to deal with concrete problems by change `version` field.

Overall, I'm with your design for the RV transferred between AS and RVPS.

Reply:
Yes, this information is indeed what the RVPS should record and maintain. My suggestions here are only for the structure transmitted by the RVPS to the AS :)